compiler/src/compiler/ast_to_hir.rs

```diff
@@ -260,6 +261,32 @@ impl<'a> Context<'a> {
         }
     }
 
+    fn lower_text(&mut self, id: Option<ast::Id>, text: &Text) -> hir::Id {
+        // TODO: Convert parts to text
+        if let [first, rest @ ..] = text.0.as_slice() {
+            let mut result = self.compile_single(first);
+            for part in rest {
+                let part_id = self.compile_single(part);
+                let concatenate = self.push(
+                    None,
+                    Expression::Builtin(BuiltinFunction::TextConcatenate),
+                    None,
+                );
+                result = self.push(
+                    None,
+                    Expression::Call {
+                        function: concatenate,
+                        arguments: vec![result.clone(), part_id],
+                    },
+                    None,
+                );
+            }
+            result
+        } else {
+            self.push(id, Expression::Text("".to_string()), None)
+        }
+    }
```

Comment:
We could remove the surrounding `if` and the slice splitting if we just always join text with the empty text. We can always add optimization passes in later compiler stages that simplify code like this in the general case.

```suggestion
    fn lower_text(&mut self, id: Option<ast::Id>, text: &Text) -> hir::Id {
        // TODO: Convert parts to text
        let mut result = self.push(id, Expression::Text("".to_string()), None);
        let concatenate = self.push(
            None,
            Expression::Builtin(BuiltinFunction::TextConcatenate),
            None,
        );
        for part in text.0 {
            let part_id = self.compile_single(part);
            result = self.push(
                None,
                Expression::Call {
                    function: concatenate,
                    arguments: vec![result.clone(), part_id],
                },
                None,
            );
        }
        result
    }
```

Reply:
This breaks use statements, because they expect a text expression, but a call gets generated instead (this somehow also results in obscure errors). I think it would be best to add a FIXME here for now and support constant expressions of any kind for use statements in the future. Otherwise we would have to prevent text interpolation in the path during parsing and combine the text parts (e.g. multiline text) here.

Reply:
Ahhh I see.

AFAIK, code such as `use "Foo"` won't actually be lowered into a use expression in the HIR. Instead, we implicitly generate a `use` function at the beginning that takes a single parameter and has a use expression inside that also contains the (hardcoded) path of the current file.

Then, this is lowered to the MIR. When optimizing the MIR, we inline this generated `use` function (using the `inline_functions_containing_use` optimization pass) and then replace the calls with the actual code of the imported modules (the `module_folding` optimization).

So, parsing this as a call shouldn't be a problem if the `constant_folding` optimization would support calls of the text concatenation built-in, which it currently doesn't. To proceed, I think it's fine to either:

- Add a todo right here.
- Improve the constant folding to handle text concatenation, which shouldn't be a lot of code.

Reply:
I improved the constant folding (bd53530) and refactored the code (0445cf7). Now `use` even works with interpolations.